ruler/ruler.py

```diff
@@ -1,91 +1,127 @@
-void_indicator = ['invert', 'void', 'white', 'v', 'w', 'false', '0', True, 0]
-minimum_indicator = ['minimum', 'minimal', 'min']
-both_indicator = ['both', 'two', '2', 2]
 
+def solid_minimum_length(
+    arr: np.ndarray,
+    phys_size: Tuple[float, ...],
+    margin_size: List[Tuple[float, float]] = ((0, 0), (0, 0))
+) -> float:
+    '''
+    Compute the minimum length scale of solid regions in a design pattern.
 
-def minimum_length(
-        arr: Iterable[np.ndarray],
-        phys_size: Iterable[np.ndarray],
-        margin_size: Iterable[np.ndarray] = np.zeros((2, 2)),
-        measure_region: str = 'min',
-        pad_mode: Optional[str] = None,
-        with_boundary: bool = False) -> Tuple[float, Optional[float]]:
+    arr: input array that represents a design pattern.
+    phys_size: physical size of the design pattern.
+    margin_size: physical size of borders that need to be discarded.
     '''
-    Compute the minimum length scale in a design pattern.
+
+    phys_size, short_entire_side, pixel_size, short_pixel_side, margin_number, dims, arr = _ruler_initialize(
+        arr, phys_size, margin_size)
+
+    if dims == 1:
+        arr = arr[margin_number[0, 0]:len(arr) - margin_number[0, 1]]
+        solid_min_length, _ = _minimum_length_1d(arr)
+        return solid_min_length * short_pixel_side
+
+    arr_shape = np.array(arr.shape)
+
+    # pad_number: numbers of pixels corresponding to the extra widths padded to the design pattern.
+    pad_number = np.ceil(arr_shape / 2).astype(int)
+    pad_width = *((int(np.ceil(s / 2)), int(np.ceil(s / 2)))
+                  for s in arr.shape),
+    arr_padded = np.pad(arr, pad_width=pad_width, mode='edge')
+    padded_shape = arr_shape + 2 * pad_number
+    padded_size = padded_shape * pixel_size
+
+    def _interior_pixel_number(diameter, arr):
+        '''
+        diameter: diameter of the structuring element, which acts like a probe
+        arr: 2d or 3d input array.
+        margin_number: numbers of pixels corresponding to the marginal widths that need to be discarded.
+        pad_number: numbers of pixels corresponding to the extra widths padded to the design pattern.
+        dual: if true, calculate the difference between opening and closing;
+              if not true, calculate the difference between opening and the input array.
+        '''
+        diff_image = abs(heaviside_open(arr, diameter, padded_size) - arr)
+        return _interior_solid_pixel_count(diff_image, margin_number,
+                                           pad_number, dims)
+
+    min_len, status = _search([short_pixel_side, short_entire_side],
+                              min(pixel_size),
+                              lambda d: _interior_pixel_number(d, arr_padded))
+
+    return min_len
+
+
+def void_minimum_length(
+    arr: np.ndarray,
+    phys_size: Tuple[float, ...],
+    margin_size: List[Tuple[float, float]] = ((0, 0), (0, 0))
```

Comment:
I think you would actually want `[(0, 0), (0, 0)]`, since your typing indicates a list of tuples. OTOH this is probably not a reasonable default value since it is fixed to two dimensions and `arr` could be 1, 2, or 3 dimensions. If you want the default behavior to be 0 margin on top and bottom of all dims of `arr`, I suggest typing this as `Optional[ ... ]`, setting the default to `None`, and handling the creation of an appropriate zero-valued `margin_size` for whatever dimension of `arr` is passed inside the function.

Reply:
> I think you would actually want [(0, 0), (0, 0)], since your typing indicates a list of tuples.

I think something like `((0,0),(0,0))` may sound better because it would follow the same syntax as `pad_width` in `numpy.pad`, as you suggested before.

> I suggest typing this as Optional[ ... ], setting the default to None

In the next version, this line will be `margin_size: Optional[Tuple[Tuple[float, float],...]] = None`.